### ISSUE TITLE
libjulia_jll: adding a `debug` variant

### ISSUE THREAD

**Issue by fingolfin** [JuliaPackaging/Yggdrasil#5787]
A bunch of our JLLs link against `libjulia`, e.g. `libcxxwrap_julia_jll` or `GAP_jll` and several more.

When debugging things, it's useful to run a debug build of Julia -- but those are linked against `libjulia-debug`, not `libjulia`. This then leads to errors when trying to load the JLLs which want to use `libjulia`, resulting in this error (see https://github.com/oscar-system/GAP.jl/issues/834):
```
julia> using GAP_jll
ERROR: InitError: could not load library "/home/tim/.julia/artifacts/e9b6449c9e90ab3718db7e31df350994ebe44003/lib/libgap.so"
libjulia.so.1: cannot open shared object file: No such file or directory
Stacktrace:
  [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
    @ Base.Libc.Libdl ./libdl.jl:117
...
```

In a Slack discussion on this,  @staticfloat suggested the following:
> Easy `.pkg` platform augmentation hook; add a `debug` tag based on whether the current Julia process is debug or not, then build both types of libraries

> Debatable PR to base: include the `debug` tag by default in `HostPlatform()` so that you don’t have to do that for future Julia builds

I'd like to work towards this, but would need some help. Hopefully with help by e.g. @giordano @staticfloat @benlorenz and perhaps others we can get there... I assume @barche might also be interested in this feature?

Here is what my (possibly incorrect) understanding is of what we need to do it, and how to get there, plus questions.

1. Modify `libjulia_jll` to also build `libjulia-debug` shared libraries
   - I hope that for this it'll be sufficient to add `julia-src-debug julia-cli-debug` to the make targets?
   - (and in theory similar for older Julia versions, but I wouldn't bother with those)
   - I don't think we'd need `debug` variants of `libjulia_jll`, though -- we can just ship both shared libraries side by side? At least I see no immediate advantage for such a split (remember, this JLL is only used for building other JLLs, users are never supposed to install or load it)
2. tweak JLLs using `libjulia_jll` (such as `GAP_jll`, `libcxxwrap_julia_jll` etc.) to offer `debug` variants (as separate artifacts) which will be built exactly like the regular ones, only they use `-ljulia-debug` instead of `-ljulia`
   - this may need some patching but it shouldn't be hard in each case; liberally applying `sed "s/-ljulia/-ljulia-debug/"` to a bunch of files may already be sufficient
   - we then add the variants to the list of platforms. How exactly? I guess with a function similar to `expand_cxxstring_abis`: it takes a list of platforms, and for each platform `p` in that list it outputs two platforms: `p` and a copy of `p` with `p["debug"] = "true"` or so set
   - or should the `debug` key in both platforms, once set to "true" and once to "false" (or perhaps `"1"` and `"0"` ???
3. Finally the packages using those JLLs need to (a) detect if Julia is running in debug mode (I assume there is a function for this, but I have not yet looked; if someone knows how to do this by heart, please comment), then (b) somehow ensure that the "debug" key in the platform is set appropriately, *before* loading the JLL...
   - how to detect if Julia is in debug mode?
   - how to ensure that the `debug` key is set appropriately in the platform when loading the JLLs? Or is that perhaps the wrong question
4. profit!

Is that about right? I'd appreciate any comments, clarifications, suggestions, help... :-)

**Comment by staticfloat** [comment]
> How exactly? I guess with a function similar to expand_cxxstring_abis: it takes a list of platforms, and for each platform p in that list it outputs two platforms: p and a copy of p with p["debug"] = "true" or so set

Yes, that's correct.  Basically write a function that takes in a list of platforms, then outputs teh same list twice, once with `debug=false`, once with `debug=true`.  Then, within your bash build script, check `if [[ "${bb_full_target}" == *debug+true* ]]` and if so, link against `libjulia-debug` instead of `libjulia`.

> how to detect if Julia is in debug mode?

```
julia> ccall(:jl_is_debugbuild, Cint, ())
0
```
This means this is NOT a debug build.

> how to ensure that the debug key is set appropriately in the platform when loading the JLLs?

This gets us into the delightful world of pkg platform augmentation hooks.  Read [this section of the Pkg docs](https://pkgdocs.julialang.org/v1/artifacts/#Extending-Platform-Selection), the very last code snippet tells you how to do this properly.

**Comment by fingolfin** [comment]
Thanks, that's all very helpful! So for the last point, one to do this:

Add a file `.pkg/platform_augmentation.jl` with content roughly like this:
```libjulia
using Libdl, Base.BinaryPlatforms
function augment_platform!(p::Platform)
    if !haskey(p, "debug")
        p["debug"] = ccall(:jl_is_debugbuild, Cint, ()) != 0
    end
    return p
end
```
Then add `.pkg/select_artifacts.jl` roughly like this (basically copied verbatim):
```julia
using TOML, Artifacts, Base.BinaryPlatforms
include("./platform_augmentation.jl")
artifacts_toml = joinpath(dirname(@__DIR__), "Artifacts.toml")

# Get "target triplet" from ARGS, if given (defaulting to the host triplet otherwise)
target_triplet = get(ARGS, 1, Base.BinaryPlatforms.host_triplet())

# Augment this platform object with any special tags we require
platform = augment_platform!(HostPlatform(parse(Platform, target_triplet)))

# Select all downloadable artifacts that match that platform
artifacts = select_downloadable_artifacts(artifacts_toml; platform)

# Output the result to `stdout` as a TOML dictionary
TOML.print(stdout, artifacts)
```

Finally access artifacts like this:
```julia
include("../.pkg/platform_augmentation.jl")

function get_artifact_dir(name)
    p = augment_platform!(HostPlatform())
    return @artifact_str(name, p)
end

function __init__()
    global my_artifact_dir = get_artifact_dir("MyArtifact")
end
```


However, the package with the artifacts is a JLL. What I don't understand is:  how could I inject such code into the JLL? Would this require changes to `JLLWrappers`?

**Comment by giordano** [comment]
Did you have a look at the CUDA, MPI and LLVM packages using the files in the [`platforms/`](https://github.com/JuliaPackaging/Yggdrasil/tree/d2cbdb2501f0e25c29e34575b302a8c4b1498737/platforms) directory?

**Comment by barche** [comment]
> I assume @barche might also be interested in this feature?

I'm willing to help make libcxxwrap-julia compatible with this, but for me personally I just build debug binaries myself when I need debug info, and use them through Overrides.toml.

**Comment by fingolfin** [comment]
@giordano no, I wasn't aware of them. Thank you for the pointer, I'll take a look now

@barche OK

**Comment by giordano** [comment]
Fixed by #5838?
